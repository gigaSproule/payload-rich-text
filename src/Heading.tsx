import type { TextData } from "./Text";
import { Text } from "./Text";
import type { Options } from "./types";

export interface HeadingData {
  type: "heading";
  children: TextData[];
  direction: "ltr" | "rtl" | null;
  format: string;
  indent: number;
  version: number;
  tag: "h1" | "h2" | "h3" | "h4" | "h5" | "h6";
}

export type Props = {
  data: HeadingData;
  options?: Options;
};

export const Heading = ({ data, options }: Props) => {
  if (data.children.length === 0) {
    return null;
  }
  const children = data.children.map((headingChild) => {
    return (
      <Text key={headingChild.text} data={headingChild} options={options} />
    );
  });
  if (options?.heading) {
    return options.heading(data, children);
  } else {
    switch (data.tag) {
      case "h1":
        return <h1>{children}</h1>;
      case "h2":
        return <h2>{children}</h2>;
      case "h3":
        return <h3>{children}</h3>;
      case "h4":
        return <h4>{children}</h4>;
      case "h5":
        return <h5>{children}</h5>;
      case "h6":
        return <h6>{children}</h6>;
    }
  }
};
