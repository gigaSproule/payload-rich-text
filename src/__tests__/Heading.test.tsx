import { describe, expect, it } from "@jest/globals";
import { render } from "@testing-library/react";
import type { Props } from "../Heading";
import { Heading } from "../Heading";
import { createHeadingData } from "./helpers/createHeadingData";

describe("Heading", () => {
  it("should not render anything if no children are provided", () => {
    const headingData: Props["data"] = createHeadingData({ children: [] });
    const { container } = render(<Heading data={headingData} />);
    expect(container).toMatchSnapshot();
  });

  (["h1", "h2", "h3", "h4", "h5", "h6"] as Props["data"]["tag"][]).forEach(
    (tag) =>
      it(`should render default ${tag} if no options provided`, () => {
        const headingData: Props["data"] = createHeadingData({
          tag: tag,
        });
        const { container } = render(<Heading data={headingData} />);
        expect(container).toMatchSnapshot();
      }),
  );

  (["h1", "h2", "h3", "h4", "h5", "h6"] as Props["data"]["tag"][]).forEach(
    (tag) =>
      it(`should render default ${tag} if options provided without heading`, () => {
        const headingData: Props["data"] = createHeadingData({
          tag: tag,
        });
        const { container } = render(
          <Heading data={headingData} options={{}} />,
        );
        expect(container).toMatchSnapshot();
      }),
  );

  it("should render custom render if options provided with heading", () => {
    const headingData: Props["data"] = createHeadingData();
    const { container } = render(
      <Heading
        data={headingData}
        options={{
          heading: (heading, children) => {
            return (
              <div>
                <span>Custom Heading - {JSON.stringify(heading)}</span>
                {children}
              </div>
            );
          },
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
