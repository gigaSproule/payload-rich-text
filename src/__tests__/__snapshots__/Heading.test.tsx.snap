// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Heading should not render anything if no children are provided 1`] = `<div />`;

exports[`Heading should render custom render if options provided with heading 1`] = `
<div>
  <div>
    <span>
      Custom Heading - 
      {"children":[{"detail":0,"format":0,"mode":"normal","style":"","text":"heading text","type":"text","version":1}],"direction":null,"format":"","indent":0,"type":"heading","version":1,"tag":"h1"}
    </span>
    heading text
  </div>
</div>
`;

exports[`Heading should render default h1 if no options provided 1`] = `
<div>
  <h1>
    heading text
  </h1>
</div>
`;

exports[`Heading should render default h1 if options provided without heading 1`] = `
<div>
  <h1>
    heading text
  </h1>
</div>
`;

exports[`Heading should render default h2 if no options provided 1`] = `
<div>
  <h2>
    heading text
  </h2>
</div>
`;

exports[`Heading should render default h2 if options provided without heading 1`] = `
<div>
  <h2>
    heading text
  </h2>
</div>
`;

exports[`Heading should render default h3 if no options provided 1`] = `
<div>
  <h3>
    heading text
  </h3>
</div>
`;

exports[`Heading should render default h3 if options provided without heading 1`] = `
<div>
  <h3>
    heading text
  </h3>
</div>
`;

exports[`Heading should render default h4 if no options provided 1`] = `
<div>
  <h4>
    heading text
  </h4>
</div>
`;

exports[`Heading should render default h4 if options provided without heading 1`] = `
<div>
  <h4>
    heading text
  </h4>
</div>
`;

exports[`Heading should render default h5 if no options provided 1`] = `
<div>
  <h5>
    heading text
  </h5>
</div>
`;

exports[`Heading should render default h5 if options provided without heading 1`] = `
<div>
  <h5>
    heading text
  </h5>
</div>
`;

exports[`Heading should render default h6 if no options provided 1`] = `
<div>
  <h6>
    heading text
  </h6>
</div>
`;

exports[`Heading should render default h6 if options provided without heading 1`] = `
<div>
  <h6>
    heading text
  </h6>
</div>
`;
